File: internal/sam/messages.go
package sam

type MsgType string

const (
	MsgReady  MsgType = "ready"
	MsgResult MsgType = "result"
)

// ReadyMsg — первое сообщение receiver'а: SAM-сессия создана,
// STREAM ACCEPT вызван, готов к входящему соединению.
type ReadyMsg struct {
	Type MsgType `json:"type"`
	Dest string  `json:"dest"` // полный base64-ключ destination (I2PAddr), не b32
}

// ResultMsg — итоговые метрики. Поля с omitempty отсутствуют
// у той стороны, которая их не измеряет.
type ResultMsg struct {
	Type         MsgType `json:"type"`
	Role         string  `json:"role"`                    // "sender" | "receiver"
	SetupMs      float64 `json:"setup_ms,omitempty"`      // sender: от NewSAM до DialI2P вернул
	TransferMs   float64 `json:"transfer_ms"`             // время передачи payload
	FirstByteMs  float64 `json:"first_byte_ms,omitempty"` // receiver: от Accept до первого байта
	GoodputMbps  float64 `json:"goodput_mbps,omitempty"`  // sender: payload/transfer_ms
	PayloadBytes int64   `json:"payload_bytes"`
	SHA256OK     bool    `json:"sha256_ok"`
	Success      bool    `json:"success"`
	Error        string  `json:"error,omitempty"`
}

File: sam3/src/messages.rs
use serde::{Deserialize, Serialize};

fn is_zero(v: &f64) -> bool {
    *v == 0.0
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum MsgType {
    Ready,
    Result,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ReadyMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub dest: String,
}

impl ReadyMsg {
    pub fn new(dest: String) -> Self {
        Self {
            msg_type: "ready".to_string(),
            dest,
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ResultMsg {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub role: String,
    #[serde(skip_serializing_if = "is_zero", default)]
    pub setup_ms: f64,
    pub transfer_ms: f64,
    #[serde(skip_serializing_if = "is_zero", default)]
    pub first_byte_ms: f64,
    #[serde(skip_serializing_if = "is_zero", default)]
    pub goodput_mbps: f64,
    pub payload_bytes: i64,
    pub sha256_ok: bool,
    pub success: bool,
    #[serde(skip_serializing_if = "String::is_empty", default)]
    pub error: String,
}

impl ResultMsg {
    pub fn new(role: &str) -> Self {
        Self {
            msg_type: "result".to_string(),
            role: role.to_string(),
            setup_ms: 0.0,
            transfer_ms: 0.0,
            first_byte_ms: 0.0,
            goodput_mbps: 0.0,
            payload_bytes: 0,
            sha256_ok: false,
            success: false,
            error: String::new(),
        }
    }
}
